File: src/Alimer/Engine/Components/MeshComponent.cs
// Copyright (c) Amer Koleci and Contributors.
// Licensed under the MIT License (MIT). See LICENSE in the repository root for more information.

using Alimer.Rendering;

namespace Alimer.Engine;

public sealed partial class MeshComponent : RenderableComponent
{
    /// Local-space bounding box.
    private BoundingBox _localBoundingBox;
    /// World-space bounding box.
    private BoundingBox _worldBoundingBox;
    /// World-space bounding box dirty flag.
    private bool _worldBoundingBoxDirty;

    public MeshComponent()
    {
    }

    public MeshComponent(Mesh mesh)
    {
        Mesh = mesh;
        _localBoundingBox = mesh.Bounds;
        _worldBoundingBoxDirty = true;
    }

    public Mesh? Mesh { get; set; }

    public IList<Material> Materials { get; } = [];
    public BoundingBox LocalBoundingBox => _localBoundingBox;

    public BoundingBox WorldBoundingBox
    {
        get
        {
            if (Entity is null)
                return _localBoundingBox;

            if (!_worldBoundingBoxDirty)
                return _worldBoundingBox;

            //OnWorldBoundingBoxUpdate();
            _worldBoundingBox = BoundingBox.Transform(_localBoundingBox, Entity!.Transform.WorldMatrix);
            _worldBoundingBoxDirty = false;
            return _worldBoundingBox;
        }
    }
}

File: Source/Alimer/Renderer/Components/MeshComponent.cpp
// Copyright (c) Amer Koleci and Contributors.
// Licensed under the MIT License (MIT). See LICENSE in the repository root for more information.

#include "Alimer/Renderer/Components/MeshComponent.h"

using namespace Alimer;

void MeshComponent::Register()
{
    auto reflection = GetTypeInfoReflection(MeshComponent::GetTypeInfoStatic());
    reflection->SetFactory<MeshComponent>();
    reflection->SetCategory("Rendering");
    reflection->SetDisplayName("Mesh Component");

    //RegisterMixedRefProperty("mesh", &MeshComponent::GetMeshAttr, &MeshComponent::SetMeshAttr, AssetRef(Mesh::GetTypeStatic()));
}

MeshComponent::MeshComponent(Mesh* mesh)
{
    _requireTransformChangeListener = true;
    SetMesh(mesh);
}

void MeshComponent::SetMesh(Mesh* value)
{
    if (_mesh == value)
        return;

    _mesh = value;
    if (value)
    {
        _materials.resize(value->GetSubmeshCount());
        //for (size_t i = 0; i < materials.size(); ++i)
        //    materials[i] = Material::DefaultMaterial();

        _localBoundingBox = value->GetBoundingBox();
    }
    else
    {
        _materials.clear();
        _localBoundingBox = {};
    }

    _worldBoundingBoxDirty = true;
}
